### PR TITLE
Code quality

### PR DESCRIPTION
**Pull request by lxgr-linux**
Restructuring for better readability, outsourcing and in general better code.
This also fixes some smaller bugs, like the "wrong hp" bug.

### CODE REVIEW

File: pokete_classes/input.py [hunk 1]
```diff
@@ -27,7 +27,8 @@ def text_input(obj, _map, name, ev, wrap_len, max_len=1000000):
             obj.rechar(hard_liner(wrap_len, name + "█"))
             _map.show()
             ev.clear()
-        elif ev.get() not in ["", "Key.enter", "exit", "Key.backspace", "Key.shift",
+        elif ev.get() not in ["", "Key.enter", "exit", "Key.backspace",
+                              "Key.shift",
                         "Key.shift_r", "Key.esc"] and len(name) < max_len:
             if ev.get() == "Key.space":
```
Comment (MaFeLP) [inline pokete_classes/input.py:33]:
```suggestion
        elif ev.get() not in ["", "Key.enter", "exit", "Key.backspace",
                              "Key.shift",
                              "Key.shift_r", "Key.esc"] and len(name) < max_len:
```

File: pokete_classes/learnattack.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                            _ev.clear()
+                            ask_ok(_ev, self.map,
+                                   f"{self.poke.name} learned \
+{attacks[new_attack]['name']}!")
+                            _ev.clear()
+                            break
+                        elif _ev.get() == "'1'":
+                            _ev.clear()
+                            Detail(self.map.height, self.map.width)\
+                                    (_ev, self.poke, False)
```
Comment (MaFeLP) [inline pokete_classes/learnattack.py:64]:
```suggestion
                            Detail(self.map.height, self.map.width)\
                                  (_ev, self.poke, False)
```

File: pokete_classes/movemap.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+        self.multitext = OutP("", state="float")
+        self.underline = se.Square("-", self.width, 1, state="float")
+        self.code_label.add(self, 0, 0)
+
+    def add_obs(self):
+        """Adds needed labels to movemap"""
+        self.underline.add(self, 0, self.height - 2)
+        self.name_label.add(self, 2, self.height - 2)
+        self.balls_label.add(self, 4 + len(self.name_label.text),
+                            self.height - 2)
```
Comment (MaFeLP) [inline pokete_classes/movemap.py:31]:
```suggestion
        self.balls_label.add(self, 4 + len(self.name_label.text),
                             self.height - 2)
```

File: pokete_classes/movemap.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        self.label_bg.resize(self.width, 1)
+        self.add_obs()
+
+    def balls_label_rechar(self, pokes):
+        """Rechars the balls label"""
+        self.balls_label.rechar("".join("-" if i >= len(pokes)
+                                or pokes[i].identifier == "__fallback__"\
+                                        else "o" if pokes[i].hp > 0
+                                        else "x"
+                                    for i in range(6)), esccode=Color.thicc)
```
Comment (MaFeLP) [inline pokete_classes/movemap.py:85]:
```suggestion
                                        for i in range(6)), esccode=Color.thicc)
```

File: pokete_classes/movemap.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                            self.height - 2)
+        self.label_bg.add(self, 0, self.height - 1)
+        self.label.add(self, 0, self.height - 1)
+
+    def text(self, x, y, arr, _ev):
+        """Shows dialog text on movemap"""
+        # This ensures the game does not crash when big
+        # chunks of text are displayed
+        for _c, i, j, _k in zip([x, y], ["x", "y"],
+                              [self.width, self.height], [17, 10]):
```
Comment (MaFeLP) [inline pokete_classes/movemap.py:40]:
```suggestion
        for _c, i, j, _k in zip([x, y], ["x", "y"],
                                [self.width, self.height], [17, 10]):
```

File: pokete_classes/moves.py [hunk 6]
```diff
@@ -84,7 +84,8 @@ def shine(self, ico=Color.thicc + Color.green + "*" + Color.reset):
 
     def downgrade(self):
         """Downgrade move"""
-        self.poke.enem.moves.shine(ico=Color.thicc + Color.red + "-" + Color.reset)
+        self.poke.enem.moves.shine(ico=Color.thicc + Color.red + "-"
+                                       + Color.reset)
```
Comment (MaFeLP) [inline pokete_classes/moves.py:88]:
```suggestion
                                   + Color.reset)
```

File: pokete_classes/roadmap.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+                      and self.sta.is_city()):
+                    return self.sta.associates[0]
+                elif (_ev.get() == "Key.enter" and not choose
+                      and self.sta.has_been_visited()):
+                    _ev.clear()
+                    p_list = ", ".join(set(self.p_d.pokes[j]["name"]
+                                        for i in self.sta.associates
+                                            for j in
+                                                i.poke_args.get("pokes", [])
+                                              + i.w_poke_args.get("pokes", [])))
```
Comment (MaFeLP) [inline pokete_classes/roadmap.py:119]:
```suggestion
                    p_list = ", ".join(set(self.p_d.pokes[j]["name"]
                                       for i in self.sta.associates
                                           for j in
                                           i.poke_args.get("pokes", [])
                                           + i.w_poke_args.get("pokes", [])))
```

File: pokete_classes/side_loops.py [hunk 8]
```diff
@@ -90,12 +92,15 @@ def __init__(self, ver, codename):
...
+            .add(self.map, int(self.map.width / 2) - 15,
+                 int(self.map.height / 2) - 4)
         se.Text(f"v{ver}", state="float").add(self.map,
-                                              int(self.map.width / 2) - 15, int(self.map.height / 2) + 2)
+                                              int(self.map.width / 2) - 15,
+                                              int(self.map.height / 2) + 2)
         se.Text(codename, state="float").add(self.map,
-                                             int(self.map.width / 2) + 14 - len(codename),
+                                             int(self.map.width / 2) + 14
+                                                - len(codename),
```
Comment (MaFeLP) [inline pokete_classes/side_loops.py:103]:
```suggestion
                                             - len(codename),
```

File: pokete_classes/types.py [hunk 9]
```diff
@@ -19,11 +19,16 @@ def __init__(self, name, effective, ineffective, color):
         self.name = name
         self.effective = effective
         self.ineffective = ineffective
-        self.color = "" if color is None else str.join("", [getattr(Color, i) for i in color])
+        self.color = "" if color is None else "".join(getattr(Color, i)
+                                                        for i in color)
```
Comment (MaFeLP) [inline pokete_classes/types.py:23]:
```suggestion
                                                      for i in color)
```

File: pokete.py [hunk 10]
```diff
@@ -196,7 +205,8 @@ def do(self, _map):
         """Interaction with the trainer"""
         if figure.has_item("shut_the_fuck_up_stone"):
             return
-        if figure.x == self.x and self.poke.hp > 0 and (self.name not in used_npcs or not settings.save_trainers):
+        if figure.x == self.x and self.poke.hp > 0 and (self.name
+                            not in used_npcs or not settings.save_trainers):
```
Comment (MaFeLP) [inline pokete.py:209]:
```suggestion
        if figure.x == self.x and self.poke.hp > 0 and (self.name
                                                        not in used_npcs or not
                                                        settings.save_trainers):
```

File: pokete.py [hunk 11]
```diff
@@ -218,18 +228,19 @@ def do(self, _map):
...
+                movemap.text(self.x, self.y, self.texts, ev)
+                winner = fight([poke for poke in figure.pokes[:6]
+                                    if poke.hp > 0][0],
                                self.poke, info={"type": "duel", "player": self})
-                movemap_text(self.x, self.y, {True: self.lose_texts,
+                movemap.text(self.x, self.y, {True: self.lose_texts,
                                               False: self.win_texts
                                                      + [" < Here u go 20$"]}
-                                              [winner == self.poke])
+                                              [winner == self.poke], ev)
```
Comment (MaFeLP) [inline pokete.py:238]:
```suggestion
                             [winner == self.poke], ev)
```

File: pokete.py [hunk 12]
```diff
@@ -218,18 +228,19 @@ def do(self, _map):
                 movemap.full_show()
                 time.sleep(0.3)
             if any(poke.hp > 0 for poke in figure.pokes[:6]):
-                movemap_text(self.x, self.y, self.texts)
-                winner = fight([poke for poke in figure.pokes[:6] if poke.hp > 0][0],
+                movemap.text(self.x, self.y, self.texts, ev)
+                winner = fight([poke for poke in figure.pokes[:6]
+                                    if poke.hp > 0][0],
```
Comment (MaFeLP) [inline pokete.py:233]:
```suggestion
                                if poke.hp > 0][0],
```

File: pokete.py [hunk 13]
```diff
@@ -244,24 +255,28 @@ def action(self, ob):
...
+\n c: Go"], ev)
         while True:
             if ev.get() == "'a'":
                 ev.clear()
                 while "__fallback__" in [p.identifier for p in figure.pokes]:
-                    figure.pokes.pop([p.identifier for p in figure.pokes].index("__fallback__"))
-                balls_label_rechar()
+                    figure.pokes.pop([p.identifier
+                                        for p in
+                                            figure.pokes].index("__fallback__"))
```
Comment (MaFeLP) [inline pokete.py:269]:
```suggestion
                    figure.pokes.pop([p.identifier for p in
                                      figure.pokes].index("__fallback__"))
```

File: pokete.py [hunk 14]
```diff
@@ -420,15 +494,16 @@ def set_player(self, player):
     def set_vars(self):
         """Updates/sets some vars"""
         for name in ["atc", "defense", "initiative"]:
-            setattr(self, name, self.lvl() + self.inf[name] + (2 if self.shiny else 0))
+            setattr(self, name, self.lvl() + self.inf[name]
+                                + (2 if self.shiny else 0))
```
Comment (MaFeLP) [inline pokete.py:498]:
```suggestion
            setattr(self, name, self.lvl() + self.inf[name]
                    + (2 if self.shiny else 0))
```

File: pokete.py [hunk 15]
```diff
@@ -450,7 +525,8 @@ def label_rechar(self):
         """Rechars the attack labels"""
         for i, atc in enumerate(self.attac_obs):
             self.atc_labels[i].rechar(f"{i + 1}: ")
-            self.atc_labels[i] += se.Text(atc.name, esccode=atc.type.color) + se.Text(f"-{atc.ap}")
+            self.atc_labels[i] += se.Text(atc.name, esccode=atc.type.color)\
+                                + se.Text(f"-{atc.ap}")
```
Comment (MaFeLP) [inline pokete.py:529]:
```suggestion
                + se.Text(f"-{atc.ap}")
```

File: pokete.py [hunk 16]
```diff
@@ -872,56 +900,31 @@ def __call__(self, p_len, label="Your full deck", in_fight=False):
...
-
     def add_all(self, pokes, init=False):
         """Adds all Poketes to the deck"""
         j = 0
         for i, poke in enumerate(pokes):
-            self.add(poke, self.map,
-                     1 if i % 2 == 0 else round(self.map.width / 2) + 1, j * 5 + 1)
+            self.add(poke, figure, self.map,
+                     1 if i % 2 == 0
+                        else round(self.map.width / 2) + 1, j * 5 + 1)
```
Comment (MaFeLP) [inline pokete.py:924]:
```suggestion
                     else round(self.map.width / 2) + 1, j * 5 + 1)
```

File: pokete.py [hunk 17]
```diff
@@ -937,7 +940,9 @@ def control(self, pokes, _ev):
                                        [-1, 1, 2, -2],
                                        [len(pokes) - 1, 0,
                                         self.index.index % 2,
-                                        [i for i in range(len(pokes)) if i % 2 == self.index.index % 2][-1]]):
+                                        [i for i in range(len(pokes))
+                                            if i % 2 ==
+                                                self.index.index % 2][-1]]):
```
Comment (MaFeLP) [inline pokete.py:945]:
```suggestion
                                            self.index.index % 2][-1]]):
```

File: pokete.py [hunk 18]
```diff
@@ -1106,14 +1004,19 @@ def __call__(self):
                                             self.map.show(init=True)
                                             break
                                         poke = figure.pokes[index]
-                                        if getattr(types, obj.attack_dict['types'][0]) in poke.types:
+                                        if getattr(types,
+                                                   obj.attack_dict['types'][0])\
+                                                        in poke.types:
```
Comment (MaFeLP) [inline pokete.py:1009]:
```suggestion
                                                in poke.types:
```

File: pokete.py [hunk 19]
```diff
@@ -1148,8 +1053,9 @@ def rem_item(self, name, items):
 
     def add(self):
         """Adds all items to the box"""
-        items = [getattr(self, i) for i in figure.inv if figure.inv[i] > 0]
-        self.box.add_c_obs([se.Text(f"{i.pretty_name}s : {figure.inv[i.name]}") for i in items])
+        items = [getattr(invitems, i) for i in figure.inv if figure.inv[i] > 0]
+        self.box.add_c_obs([se.Text(f"{i.pretty_name}s : {figure.inv[i.name]}")
+                                for i in items])
```
Comment (MaFeLP) [inline pokete.py:1058]:
```suggestion
                            for i in items])
```

File: pokete.py [hunk 20]
```diff
@@ -1361,7 +1196,9 @@ def __call__(self, pokes):
         p_dict = {i[1]: i[-1] for i in
                   sorted([(pokes[j]["types"][0], j, pokes[j])
                           for j in list(pokes)[1:]])}
-        self.obs = [se.Text(f"{i + 1} {p_dict[poke]['name'] if poke in caught_poketes else '???'}", state="float")
+        self.obs = [se.Text(f"{i + 1} \
+{p_dict[poke]['name'] if poke in figure.caught_pokes else '???'}",
+                        state="float")
```
Comment (MaFeLP) [inline pokete.py:1201]:
```suggestion
                    state="float")
```

File: pokete.py [hunk 21]
```diff
@@ -1499,8 +1272,10 @@ def save():
         "inv": figure.inv,
         "money": figure.get_money(),
         "settings": settings.dict(),
-        "caught_poketes": list(dict.fromkeys(caught_poketes + [i.identifier for i in figure.pokes])),
-        "visited_maps": visited_maps,
+        "caught_poketes": list(dict.fromkeys(figure.caught_pokes
+                                             + [i.identifier
+                                                    for i in figure.pokes])),
```
Comment (MaFeLP) [inline pokete.py:1277]:
```suggestion
                                                for i in figure.pokes])),
```

File: pokete.py [hunk 22]
```diff
@@ -1841,10 +1402,13 @@ def playmap_21():
...
         """Cave animation"""
-        for obj in ob_maps["playmap_7"].inner_walls.obs + ob_maps["playmap_7"].trainers + [
-            getattr(ob_maps["playmap_7"], i) for i in p_data.map_data["playmap_7"]["balls"] if
-                "playmap_7." + i not in used_npcs or not save_trainers]:
-            if obj.added and math.sqrt((obj.y - figure.y) ** 2 + (obj.x - figure.x) ** 2) <= 3:
+        for obj in ob_maps["playmap_7"].inner_walls.obs\
+                   + ob_maps["playmap_7"].trainers\
+                   + [getattr(ob_maps["playmap_7"], i)
+                    for i in p_data.map_data["playmap_7"]["balls"] if
+                        "playmap_7." + i not in used_npcs or not save_trainers]:
```
Comment (MaFeLP) [inline pokete.py:1409]:
```suggestion
            + ob_maps["playmap_7"].trainers\
            + [getattr(ob_maps["playmap_7"], i)
               for i in p_data.map_data["playmap_7"]["balls"] if
               "playmap_7." + i not in used_npcs or not save_trainers]:
```

File: pokete.py [hunk 23]
```diff
@@ -1947,7 +1516,8 @@ def swap_poke():
     if (index := deck(6, "Your deck", True)) is None:
         return
     if do:
-        with InfoBox(f"Hostname: {socket.gethostname()}\nWaiting...", _map=movemap):
+        with InfoBox(f"Hostname: {socket.gethostname()}\nWaiting...",
+                _map=movemap):
```
Comment (MaFeLP) [inline pokete.py:1520]:
```suggestion
                     _map=movemap):
```

**Review by MaFeLP — CHANGES_REQUESTED** [review]

File: pokete_classes/input.py (hunk 1 above)
Comment (lxgr-linux) [reply pokete_classes/input.py:33]: where did the `if ev.get() == "Key.space":` go?

File: pokete_classes/fightmap.py [hunk 24]
```diff
@@ -0,0 +1,421 @@
...
+        self.outp.add(self, 1, self.height - 4)
+        self.e_underline.add(self, 1, 4)
+        self.e_sideline.add(self, len(self.e_underline.text), 1)
+        self.p_upperline.add(self,
+                                 self.width - 1 - len(self.p_upperline.text),
+                                 self.height - 10)
+        self.frame_big.add(self, 0, 0)
+        self.p_sideline.add(self,
+                                self.width - 1 - len(self.p_upperline.text),
+                                self.height - 9)
```
[the remark this answers is not in the document]
Comment (lxgr-linux) [reply pokete_classes/fightmap.py:66]: where did line 61-63 go?

File: pokete_classes/fightmap.py [hunk 25]
```diff
@@ -0,0 +1,421 @@
...
+                                        weights=[i.ap * ((1.5
+                                                          if enem.type.name in
+                                                                i.type.effective
+                                                          else 0.5
+                                                          if enem.type.name in
+                                                                i.type.ineffective
+                                                          else 1)
+                                                         if info["type"] == "duel"
+                                                         else 1)
+                                                 for i in obj.attac_obs])[0]
```
[the remark this answers is not in the document]
Comment (lxgr-linux) [reply pokete_classes/fightmap.py:291]: what happened to line 293?

**Review by MaFeLP — APPROVED** [review]

**Pull request merged**